### PR TITLE
Capstone stages

### PR DESCRIPTION
**Pull request by codeluggage**
# Feature: Capstone levels with multiple stages

We need a type of capstone level that can have multiple completions. This means that as the student is playing, they are effectively on a stage of the capstone level. 

For each stage that is completed, a cinematic will be shown. Once the cinematic has been watched, the level can be continued - now at the next stage. 

# How to test/review this PR

Due to the questions around deploying this safely to Next for testing, this needs to be reviewed locally. That means creating a new level to serve as the capstone level, in order to test the functionality. This is fairly involved, so it is explained in [this Slab doc](https://codecombat.slab.com/posts/capstone-levels-guide-rcosczss) as a general guide for Capstone Levels.

# Implementation

- `Level` now has a new property called `additionalGoals`, deciding how many stages are possible to complete in the level.
- `LevelSession` now has a new state property called `capstoneStage`, keeping track of what stage the student has reached.

When a student reaches victory for their level session, we check if it is a capstone level. If it is, we add the capstone stage to the level session - marking that they have completed stage 1.

The victory flow is retained, but actually marking the level as completed is stopped at the `LevelBus` step for `onVictory` - preventing the `complete` state to be set for the level session.

A new progression modal is shown when progressing through the stages, a new victory modal is shown when the capstone level has been fully completed. The contents here is TBD - two new models have been created, but right now a simple game dev modal is being used.

# Deploy

This needs to be deployed only after https://github.com/codecombat/codecombat/pull/5383 merges, so schemas are in order.

This feature is effectively "gated" as it is only in effect once a level has `additionalGoals`. Without this new property, the `capstoneStage` on the level session will not progress. 

As such, this can be deployed without affecting anything in production as preparation for the upcoming interactives work. 

# Questions

- The tests cover the `GoalManager` well, but not the flow through the system. I could not find code that would test this - is this something we would test with smoke tests, because of the multiple steps? 
- Are the tests readable? World generation will reset goal states and so many events typically need to be sent each step of the test. Would it be more clear with less `expect()` calls?

### CODE REVIEW

File: app/lib/LevelBus.coffee [hunk 1]
```diff
@@ -198,8 +198,9 @@ module.exports = class LevelBus extends Bus
     @changedSessionProperties.state = true
     @saveSession()
 
-  onVictory: ->
+  onVictory: (e) ->
     return unless @onPoint()
+    return if e and e.capstoneVictory
```
Comment (codeluggage) [inline app/lib/LevelBus.coffee]: Is there a better way to handle this "done but not really done" concept? 

When capstone level goals are being completed, it should behave like the level is being completed in _most ways_. However, the level is not truly complete until all the additional goals have been completed.

File: app/schemas/models/level.coffee [hunk 2]
```diff
@@ -366,7 +366,17 @@ _.extend LevelSchema.properties,
...
-
+  additionalGoals: c.array { title: 'Additional Goals', description: 'Goals that are added after the first regular goals are completed' }, {
+    type: 'object',
+    title: 'Goals',
+    description: 'Goals for this stage',
+    minItems: 1,
+    uniqueItems: true,
+    properties: {
+      stage: { type: 'integer', title: 'Goal Stage', description: 'Which stage these additional goals are for (2 and onwards)' },
+      goals: c.array { title: 'Goals', description: 'An array of goals which are visible to the player and can trigger scripts.' }, GoalSchema
```
Comment (codeluggage) [inline app/schemas/models/level.coffee]: Will we ever have optional capstone level goals? If not, does it still make sense to use `GoalSchema` here?

File: app/lib/world/GoalManager.coffee [hunk 3]
```diff
@@ -96,6 +105,37 @@ module.exports = class GoalManager extends CocoClass
...
+    state = session.get('state') ? {}
+    if not state.capstoneStage
+      state.capstoneStage = 1 # Increase from zero
+    else
+      state.capstoneStage += 1
+    goalsAdded = false
+    _.forEach(additionalGoals, (stageGoals) =>
+      if stageGoals.stage == state.capstoneStage
+        _.forEach(stageGoals.goals, (goal) =>
+          goalsAdded = true
```
Comment (codeluggage) [inline app/lib/world/GoalManager.coffee]: Could this increase the capstone level when an old goal is incomplete, then complete again?

File: app/views/play/level/LevelGoalsView.coffee [hunk 4]
```diff
@@ -53,7 +53,7 @@ module.exports = class LevelGoalsView extends CocoView
     @previousGoalStatus ?= {}
     @succeeded = e.overallStatus is 'success'
     for goal in e.goals
-      state = e.goalStates[goal.id]
+      state = e.goalStates[goal.id] ? 'incomplete'
```
Comment (codeluggage) [inline app/views/play/level/LevelGoalsView.coffee]: Make sure this is needed after `GoalManager` changes for `headless` option.

Comment (codeluggage) [reply app/views/play/level/LevelGoalsView.coffee]: For new Ozaria levels, this could be useful to always show "incomplete" for new levels with no status.

File: app/views/play/level/PlayLevelView.coffee [hunk 5]
```diff
@@ -668,6 +675,18 @@ module.exports = class PlayLevelView extends RootView
       ModalClass = CourseVictoryModal
       options.courseInstanceID = utils.getQueryVariable('course-instance') or utils.getQueryVariable('league')
     ModalClass = PicoCTFVictoryModal if window.serverConfig.picoCTF
+
+    # Progress through the capstone stages
+    if additionalGoals
+      options.capstoneStage = @session.capstoneStage
```
Comment (shubhi1092) [inline app/views/play/level/PlayLevelView.coffee]: `@session.state.capstoneStage` ?

File: app/views/play/level/PlayLevelView.coffee [hunk 6]
```diff
@@ -808,8 +827,12 @@ module.exports = class PlayLevelView extends RootView
     return if @destroyed
     Backbone.Mediator.publish 'level:set-time', ratio: 1
     return if @level.hasLocalChanges()  # Don't award achievements when beating level changed in level editor
-    if @goalManager.checkOverallStatus() is 'success'
-      showModalFn = -> Backbone.Mediator.publish 'level:show-victory', showModal: true
+    state = @session.get('state') ? {}
+    options = { showModal: true }
+    if @goalManager.finishLevel()
+      if state.capstoneStage and @goalManager.getRemainingGoals().length == 0
```
Comment (shubhi1092) [inline app/views/play/level/PlayLevelView.coffee]: Do we still set `state.complete = true` in the session for capstone levels?

File: app/lib/LevelBus.coffee (hunk 1 above)
Comment (shubhi1092) [reply app/lib/LevelBus.coffee]: Do we not want to set complete = true on full completion of the capstone level?

File: app/lib/world/GoalManager.coffee (hunk 3 above)
Comment (shubhi1092) [reply app/lib/world/GoalManager.coffee]: Sorry I dont understand this question. Also can you explain why we need `goalsAdded` flag?

File: app/views/play/level/LevelGoalsView.coffee [hunk 7]
```diff
@@ -53,7 +53,7 @@ module.exports = class LevelGoalsView extends CocoView
     @previousGoalStatus ?= {}
     @succeeded = e.overallStatus is 'success'
     for goal in e.goals
-      state = e.goalStates[goal.id]
+      state = e.goalStates[goal.id] ? { status: 'incomplete' }
```
Comment (shubhi1092) [inline app/views/play/level/LevelGoalsView.coffee]: Why is this change required?

File: app/views/play/level/PlayLevelView.coffee [hunk 8]
```diff
@@ -47,6 +47,9 @@ CourseVictoryModal = require './modal/CourseVictoryModal'
 PicoCTFVictoryModal = require './modal/PicoCTFVictoryModal'
```
Comment (shubhi1092) [inline app/views/play/level/PlayLevelView.coffee]: We should do changes related to `playLevelView` in the new file which is in ozaria folder: https://github.com/codecombat/codecombat/blob/master/ozaria/site/views/play/level/PlayLevelView.js

File: app/views/play/level/PlayLevelView.coffee [hunk 9]
```diff
@@ -668,6 +675,18 @@ module.exports = class PlayLevelView extends RootView
...
       options.courseInstanceID = utils.getQueryVariable('course-instance') or utils.getQueryVariable('league')
     ModalClass = PicoCTFVictoryModal if window.serverConfig.picoCTF
+
+    # Progress through the capstone stages
+    if additionalGoals
+      options.capstoneStage = @session.capstoneStage
+      options.remainingGoals = @goalManager.getRemainingGoals()
+      if options.remainingGoals.length > 0
+#        ModalClass = CapstoneProgressModal
+        ModalClass = GameDevVictoryModal
```
Comment (shubhi1092) [inline app/views/play/level/PlayLevelView.coffee]: Is this modal added just for testing?

**Comment by shubhi1092** [comment]
Did not test locally, but added some questions to the code changes, mainly why do we not set `state.complete` for capstone levels?
Also is there a requirement to have the 2 new victory modals, or is it just added for testing purposes?

Below items remaining to code review:
- New victory modals
- Test cases

File: app/views/play/level/LevelGoalsView.coffee (hunk 7 above)
Comment (codeluggage) [reply app/views/play/level/LevelGoalsView.coffee]: I will review this again after more modifications to the level loading and goal manager were done. Previously, there would be scenarios where goals were not yet loaded, and it would fail here.

File: app/views/play/level/PlayLevelView.coffee (hunk 8 above)
Comment (codeluggage) [reply app/views/play/level/PlayLevelView.coffee]: I will do this! There seems to be a complication here though, where game dev levels are currently having an issue for the new `playLevelView`.

File: app/views/play/level/PlayLevelView.coffee (hunk 9 above)
Comment (codeluggage) [reply app/views/play/level/PlayLevelView.coffee]: Yes - the UX/UI and progression is still unknown, so I just used a simple dialog for the time being. The two new dialogs may or may not be used depending on which direction UX/UI take.

Comment (shubhi1092) [reply app/views/play/level/PlayLevelView.coffee]: Got it, so are we planning to remove them before merging to master? We can add a comment or a console message as a placeholder until we know more (and maybe a wrap-up task for later when we know more)

File: app/lib/world/GoalManager.coffee (hunk 3 above)
Comment (codeluggage) [reply app/lib/world/GoalManager.coffee]: This was added as the loading of additional goals for `GoalManager`'s was more in flux - will loop back over this to verify that it is not needed 👍

File: app/lib/LevelBus.coffee (hunk 1 above)
Comment (codeluggage) [reply app/lib/LevelBus.coffee]: This will be set to `complete = true` when we don't pass in this option. That is only meant to happen when all the goals are complete (was missing, is in now).

File: app/views/play/level/PlayLevelView.coffee (hunk 6 above)
Comment (codeluggage) [reply app/views/play/level/PlayLevelView.coffee]: Updated - thank you for catching this!

File: app/views/play/level/LevelGoalsView.coffee (hunk 7 above)
Comment (shubhi1092) [reply app/views/play/level/LevelGoalsView.coffee]: We want to make sure that this does not affect the existing levels

File: ozaria/site/views/play/level/PlayLevelView.js [hunk 10]
```diff
@@ -1394,13 +1419,23 @@ class PlayLevelView extends RootView {
...
-    } // Don't award achievements when beating level changed in level editor
-    if (this.goalManager.checkOverallStatus() === 'success') {
+    }
+
+    const options = { showModal: true }
+    const state = this.session.get('state')
+
+    if (this.goalManager.finishLevel()) {
+      if (state && state.capstoneStage && this.goalManager.getRemainingGoals().length === 0) {
+        options.capstoneInProgress = true
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/PlayLevelView.js]: I believe setting this would still not set `complete: true` in the overall state in `LevelBus` `onVictory` method - is that right?

File: ozaria/site/views/play/level/PlayLevelView.js [hunk 11]
```diff
@@ -1394,13 +1419,23 @@ class PlayLevelView extends RootView {
...
       return
-    } // Don't award achievements when beating level changed in level editor
-    if (this.goalManager.checkOverallStatus() === 'success') {
+    }
+
+    const options = { showModal: true }
+    const state = this.session.get('state')
+
+    if (this.goalManager.finishLevel()) {
+      if (state && state.capstoneStage && this.goalManager.getRemainingGoals().length === 0) {
```
Comment (shubhi1092) [inline ozaria/site/views/play/level/PlayLevelView.js]: Does this mean the capstone is still in progress or fully complete? Where do we publish show-victory event without `capstoneInProgress` such that overall level session state is set as `complete: true`

Comment (codeluggage) [reply ozaria/site/views/play/level/PlayLevelView.js]: It means that the capstone is still in progress. When there is no `capstoneInProgress`, it is done and goes through the regular path: setting `complete: true`. 

Or said another way, as long as `capstoneInProgress` is passed along in the options, the level is not fully complete.

Comment (shubhi1092) [reply ozaria/site/views/play/level/PlayLevelView.js]: I am not sure if I understand why the length of the remaining goals is 0 but the level is still in progress?

Comment (codeluggage) [reply ozaria/site/views/play/level/PlayLevelView.js]: You're completely right - I changed this on the other machine and missed pushing it up. Updated now.

File: app/lib/world/GoalManager.coffee [hunk 12]
```diff
@@ -96,6 +105,40 @@ module.exports = class GoalManager extends CocoClass
...
     @notifyGoalChanges()
 
+  # Adds any goals for the current capstoneStage
+  # Returns the current capstoneStage
+  addAdditionalGoals: (session, additionalGoals) ->
+    state = session.get('state') ? {}
+    if not state.capstoneStage
+      # In daily speak, we think of initial goals as stage 1 and additional goals
+      # as stage 2 and above. That is why we are starting from 2.
+      state.capstoneStage = 2
```
Comment (shubhi1092) [inline app/lib/world/GoalManager.coffee]: It seems we are thinking of capstoneStage as the current stage and not the stage completed(I was assuming this to be stage completed until now :) ).
So in this case, the stage completed = stage.capstoneStage - 1, is that right? (Want to clarify since the unit map next level logic will depend on this)

File: app/lib/world/GoalManager.coffee [hunk 13]
```diff
@@ -269,6 +312,10 @@ module.exports = class GoalManager extends CocoClass
 
   setGoalState: (goalID, status) ->
     state = @goalStates[goalID]
+    if not state
+      console.error('Failed to set goal state, possibly due to capstone stages for non-ozaria play')
+      return
```
Comment (shubhi1092) [inline app/lib/world/GoalManager.coffee]: Why is this required?

File: app/schemas/models/level.coffee [hunk 14]
```diff
@@ -393,7 +393,7 @@ _.extend LevelSchema.properties,
     minItems: 1,
     uniqueItems: true,
     properties: {
-      stage: { type: 'integer', title: 'Goal Stage', description: 'Which stage these additional goals are for (2 and onwards)' },
+      stage: { type: 'integer', minimum: 2, title: 'Goal Stage', description: 'Which stage these additional goals are for (2 and onwards)' },
```
Comment (shubhi1092) [inline app/schemas/models/level.coffee]: 👍

File: ozaria/site/templates/play/play-level-view.pug [hunk 15]
```diff
@@ -7,7 +7,7 @@ if view.showAds()
     script.
       (adsbygoogle = window.adsbygoogle || []).push({});
 
-//- include game-dev-goals.jade
+include game-dev-goals
```
Comment (shubhi1092) [inline ozaria/site/templates/play/play-level-view.pug]: Was this the reason game dev levels were not working?

File: app/lib/world/GoalManager.coffee (hunk 12 above)
Comment (shubhi1092) [reply app/lib/world/GoalManager.coffee]: Also this means that at the end of the capstone level's last stage, capstoneStage will increment to a value for which we ideally don't have any goals, but `capstoneStage - 1 = level's last stage`, right?
Can we add this as a comment somewhere here to make it clear.

File: test/app/lib/world/GoalManager.spec.coffee [hunk 16]
```diff
@@ -8,6 +9,35 @@ describe('GoalManager', ->
   stayMapGoal =  {name: 'Stay here', keepFromLeavingOffSide: {who: 'Yall'}, id: 'id'}
   getItemGoal = {name: 'Mine', getItem: {who: 'Grabby', itemID: 'Sandwich'}, id: 'id'}
   keepItemGoal = {name: 'Not Yours', keepFromGettingItem: {who: 'Grabby', itemID: 'Sandwich'}, id: 'id'}
+  # Very simple session mocking
+  session = {
```
Comment (shubhi1092) [inline test/app/lib/world/GoalManager.spec.coffee]: Nitpick: You can also use utility method to create level session - `factories.makeLevelSession` (https://github.com/codecombat/codecombat/blob/master/test/app/factories.coffee)

File: app/lib/world/GoalManager.coffee [hunk 17]
```diff
@@ -96,6 +105,40 @@ module.exports = class GoalManager extends CocoClass
       @goalStates[goalID] = goalState
     @notifyGoalChanges()
 
+  # Adds any goals for the current capstoneStage
+  # Returns the current capstoneStage
+  addAdditionalGoals: (session, additionalGoals) ->
+    state = session.get('state') ? {}
+    if not state.capstoneStage
+      # In daily speak, we think of initial goals as stage 1 and additional goals
+      # as stage 2 and above. That is why we are starting from 2.
```
Comment (shubhi1092) [inline app/lib/world/GoalManager.coffee]: Will `state.capstoneStage` be equal to 1 or undefined when the user is playing for initial goals?

File: ozaria/site/templates/play/play-level-view.pug (hunk 15 above)
Comment (AndrewJakubowicz) [reply ozaria/site/templates/play/play-level-view.pug]: Yeah. I wonder why it's showing up as a diff.

File: app/views/play/level/PlayLevelView.coffee (hunk 9 above)
Comment (codeluggage) [reply app/views/play/level/PlayLevelView.coffee]: This has been updated with progression to a cinematic place holder page and back again to keep playing the level 👍

**Comment by codeluggage** [comment]
Updated PR with modal changes and a place holder cinematic page, so the to-from interaction works. This simplifies some of the need for the guards in the code against goal states. 

Build is failing due to the missing font in bower.

File: app/lib/world/GoalManager.coffee (hunk 17 above)
Comment (codeluggage) [reply app/lib/world/GoalManager.coffee]: Yes, I agree - will look at adding it to make it easier to grok the code when reading 👍

File: ozaria/site/views/play/level/modal/CapstoneProgressModal.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+      }
+    }
+  }
+
+  destroy () {
+    if (this.onDestroy) {
+      this.onDestroy()
+    }
+
+    this.goToCinematic()
```
Comment (AndrewJakubowicz) [inline ozaria/site/views/play/level/modal/CapstoneProgressModal.js]: Is this a placeholder for navigation in the future. If so, should this be in the destroy method?

I can see that this link is also calculated within the `CapstoneProgressComponent`. Would it be advantageous to pass in the `goToCinematic` function as a prop.

Comment (AndrewJakubowicz) [reply ozaria/site/views/play/level/modal/CapstoneProgressModal.js]: Currently navigation appears to occur like so.

User gets the `CapstoneProgressComponent.vue`. They click the "Onwards" anchor.
This either routes them to the url. Or in the destruction of the page the `goToCinematic` method is called routing them.

We should prefer one or the other unless there is a reason for both.

File: app/views/core/ModalComponent.coffee [hunk 19]
```diff
@@ -8,6 +8,10 @@ module.exports = class ModalComponent extends ModalView
   propsData: null
 
   afterRender: ->
+    console.log('CAPSTONE: in afterRender')
```
Comment (AndrewJakubowicz) [inline app/views/core/ModalComponent.coffee]: Accidentally kept in?

File: app/lib/world/GoalManager.coffee (hunk 13 above)
Comment (codeluggage) [reply app/lib/world/GoalManager.coffee]: During the development of this feature, the additional goals were loaded in immediately upon completing the last additional goal of the stage. This caused challenges because the state was not propagated throughout the entire state of the website. 

These guards were added to avoid other areas of the code trying to look up the state that didn't exist yet. 

As the modals now always redirect to the cinematic placeholder (and the cinematic in the future), this is no longer an issue and these can be cleaned up 👍

File: ozaria/site/templates/play/play-level-view.pug (hunk 15 above)
Comment (codeluggage) [reply ozaria/site/templates/play/play-level-view.pug]: Will rebase again and that should clear this up 👍

File: ozaria/site/views/play/level/modal/CapstoneProgressModal.js (hunk 18 above)
Comment (codeluggage) [reply ozaria/site/views/play/level/modal/CapstoneProgressModal.js]: There is a bad state the level can get into where the state of goals are not properly propagated everywhere. This happens if goals progress while staying on the same page. 

To enforce that the user always goes to the cinematic before coming back, it's in the `destroy()`. But yes! This could be passed as a prop 👍

**Comment by codeluggage** [comment]
@AndrewJakubowicz I didn't properly follow up on your question - yes, this would be done after this ships, as part of integrating with unit map and general flow from capstone level to cinematic. Having spent some time with the cinematic parts of the code, I believe it should be straight forward when we have the UX for it. Simply use the place holder flow that is in place, and show the cinematic as intended by UX.

File: app/lib/world/GoalManager.coffee (hunk 13 above)
Comment (codeluggage) [reply app/lib/world/GoalManager.coffee]: There are still cases where this part of the code would be hit as part of updates before transitioning out of the play view, where it is useful to have graceful handling instead of erroring because the state can't be found.

File: app/views/play/level/LevelGoalsView.coffee [hunk 20]
```diff
@@ -53,7 +53,7 @@ module.exports = class LevelGoalsView extends CocoView
     @previousGoalStatus ?= {}
     @succeeded = e.overallStatus is 'success'
     for goal in e.goals
-      state = e.goalStates[goal.id]
+      state = e.goalStates[goal.id] or { status: 'incomplete' }
```
Comment (codeluggage) [inline app/views/play/level/LevelGoalsView.coffee]: Leaving this in as a way to avoid errors when the play view is in a transitioning state as the session is updated and the player has not moved on to the cinematic yet.
